### PR TITLE
Wireup date picker

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

PR on top of https://github.com/cowprotocol/cowswap/pull/1922

- Wiring up the picker with state
- Added error msgs for when:
  - `< min date
  - `> max date
  - Unable to parse inserted date
- Deadline will only be updated if valid, even if input validation/picker is bypassed

⚠️ Error msg is in desperate need of styling. Will delegate that to Michel ;)

Screenshot 2023-01-27 at 15 55 52

# To Test

1. On limit orders, open the date picker and select `custom`
2. Play around with the custom date
3. Chose a date < 30min from now
4. Chose a date > 6 months from now

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @alfetopito , just wondering why the min date can't be less than 30 min? 
Currently, I can place orders with 2 minutes duration (not less)

**Comment by elena-zh** [comment]
And a few comments: 
1. I think that it would be great to make the warning date dynamic (as time is going) 
image
2. It would be great to see the default time according to a user's time zone :) 
image
3. It would be great to open the calendar by clicking on the line, not only on the icon 
image

**Comment by alfetopito** [comment]
> Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)

Well, I honestly don't know. This was the default already.

I assume that if you want to use less you can use the pre-defined values? 🤷

**Comment by alfetopito** [comment]
> And a few comments:
> 
>     1. I think that it would be great to make the warning date dynamic (as time is going)
>        image

That might be problematic as it can cause infinite re-renders: time changes at every moment, so we'd have to be constantly re-rendering.
Of course, I could set it to update every minute or so, but this is cleared everytime the modal is re-opened, so I don't see it as a big deal

>     2. It would be great to see the default time according to a user's time zone :)
>        image

Hmmm, interesting. It was supposed to do that already, so much so that the error is in your locale. Will (try to) fix it
 
>     3. It would be great to open the calendar by clicking on the line, not only on the icon
>        image

That's a bit on the browser's court. Every browser handles it differently.
@fairlighteth do you think we could do something about this?

>     4. And it would be nice not to display the error on opening. Only after user's input (if it is not valid one)

Will fix it

**Comment by alfetopito** [comment]
1. It wasn't working as I explained. Now it'll be updated every time the `custom` modal is opened
2. Not fixed yet!
4. Fixed

**Comment by elena-zh** [comment]
> > Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)
> 
> Well, I honestly don't know. This was the default already.
> 
> I assume that if you want to use less you can use the pre-defined values? 🤷

I would align it with BE limitations. Currently, when I create a limit order with expiry value < 2 min, it fails (I reported it in #1417 , case 3). The save min expiration time (2 min) we use for Market orders.

**Comment by elena-zh** [comment]
Hey @alfetopito , cases 1, 4 work for me!
However, I have found 1 more issue: 'Clear' button is not working ;(
Can we do anything with it or we can't control its behavior?
not working

**Comment by elena-zh** [comment]
> > Hey @alfetopito , just wondering why the min date can't be less than 30 min? Currently, I can place orders with 2 minutes duration (not less)
> 
> Well, I honestly don't know. This was the default already.
> 
> I assume that if you want to use less you can use the pre-defined values? 🤷

Hey @alfetopito , according to the [discussion](https://cowservices.slack.com/archives/C044B50A02F/p1675087132578169), 10 min should be set as min deadline

**Comment by alfetopito** [comment]
@elena-zh 

1. Set deadline to 10min
2. Clear button will reset the date to the min
3. And now the timezone thing is respected (as far as I tested).
But do let me know if there's something still not working.

As for the points in https://github.com/cowprotocol/cowswap/issues/1417

2. Up to the browser. Even if a bad time is picked, the validation shouldn't allow it to be saved
3. Should be addressed here
4. This is also up to the browser unfortunately
5. Same as `2`
6. Same as `2`

File: src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:185]: I think it makes sense to move all utility functions to separated file, like a `DeadlineSelector/utils.ts`

File: src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:59]:
```typescript
const {timeZone} = Intl.DateTimeFormat().resolvedOptions()
```

File: src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:69]: Should we use Transloco here?

File: src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx
Comment (shoom3301) [inline src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:171]: Should we use Transloco here?

**Comment by elena-zh** [comment]
As a nitpick, I'd add an extra padding here:
image

File: src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx
Comment (alfetopito) [reply src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:171]: What's that?

Comment (alfetopito) [reply src/cow-react/modules/limitOrders/pure/DeadlineSelector/index.tsx:171]: Oh you mean the `<Trans>` component?

**Comment by alfetopito** [comment]
> Hey @alfetopito , great job!!! Just 2 issues:
> 
>     1. Reset button is not working in iOS (it does not reset the date to min allowed):
>        image

Fixed.
Could not make the date picker respect the min date, though.
The error message will have to do.

> 
>     2. Previously specified date is saved when switch to custom option:
> 
> 
>     * Select min required Custom date
> 
>     * Then change it to any preselected value (5 min), for example
> 
>     * Wait for at least 1 minute
> 
>     * Select Custom option once again: you will see the warning on opening, the previous value is saved.
>       Can we please reset the date/time to the min required in this case?
>       This is important due to in FF you can't change the time, so the only option to avoid warning is to select the next day.

Fixed. 
Now if the previously selected date is older than the current min date, min date will be used when the modal is opened

> 
>     3. One more usability issue with FF is that the warning appears when I select max available date:
>        image
>        But the condition is met, so I as a user don't understand why the warning is displayed.
>        image

That's the expected behaviour IMO.
The date is what it says is allowed, you are right. But you also need to take the time into account.
When you adjust the time the warning should go away.

As for the UI fixes, I expect Michel will make it look much better, not just with padding. I'll leave it up to him once this is approved.

**Review by shoom3301 — APPROVED** [review]

**Comment by elena-zh** [comment]
> That's the expected behaviour IMO.
> The date is what it says is allowed, you are right. But you also need to take the time into account.
> When you adjust the time the warning should go away.

Yes, but I described the issue for FF (Firefox): you can't change time there. For all the rest browsers it works as you described.

**Comment by elena-zh** [comment]
I have created #1964 issue for the UI issues. 
All the rest changes (besides this [note](https://github.com/cowprotocol/cowswap/pull/1932#issuecomment-1412392595)) LGTM!

**Review by elena-zh — APPROVED** [review]

**Pull request merged**
